### DISCUSSION TITLE
Non overlapping target and source

### DISCUSSION THREAD

**Discussion by netfenceventu** [psy0rz/zfs_autobackup#124] [Q&A]
Hello.

I've got the following situation:
_ snapshots and backups are "separated" (as run from different scripts);
_ cron periodically runs: "/usr/local/bin/zfs-autobackup --keep-source 1d2d auto_vm"
_ as can be seen, retention is intentionally short due to very tight space constraints;
_ at the user's request, a backup script is run, which basically does "zfs-autobackup --keep-source 10,1h1d,1d1w,1w1m,1m1y --keep-target 10,1h1d,1d1w,1w1m,1m1y --destroy-missing 360d --no-holds --no-snapshot --rollback --destroy-incompatible --verbose auto_vm backupdisk";
_ this can happen every few days, weeks, or months;
_ so all the snapshots on backupdisk are long gone from the source pool;
_ the backup fails with "backupdisk/vm/sql/disk0: Destroy missing: has no snapshots made by us. (please destroy manually)";
_ this is not strictly true, as these target snapshots *have* been made by zfs_autobackup, but I guess the lack of a common snapshot leads to this.

Ideally I'd like to add the new snapshots without deleting the old ones, unless they expired as per "--keep-target". Is this possible at all?
Alternatively I'd like to delete the old snapshots, before sending the new ones.
Can zfs_autobackup do this?

Alternatively, I came up with a few ideas:
_ delete all target snapshots before running zfs_autobackup (but I'd rather keep them);
_ rename target snapshots before running zfs_autobackup (but then I guess I would need to prune them myself in some way, wouldn't I?).

Any comment on these approach or any better idea?

Thanks.

**Comment by psy0rz** [comment]
remove the --no-holds.

this way the snapshot job knows which is the common snapshot, so it wont delete it.

**Comment by netfenceventu** [comment]
I tried your suggestion, but, as I expected, it did not work.
I'm not sure I understand it correctly, though.

The "snapshot" job is the one who removes old snapshots from the source and it did not have "--no-holds".
The "backup" job runs when the common snapshot is already gone: removing "--no-holds" from it didn't make any difference.

Did I understand it wrong?



In any case, I have no space on the source to keep the latest common snapshot for too long, so I need a way to run normally if such snapshot exists or start from scratch otherwise.
Is this possisble?



TIA

**Reply by psy0rz** [reply]
> I tried your suggestion, but, as I expected, it did not work. I'm not sure I understand it correctly, though.
> 
> The "snapshot" job is the one who removes old snapshots from the source and it did not have "--no-holds". The "backup" job runs when the common snapshot is already gone: removing "--no-holds" from it didn't make any difference.
> 
> Did I understand it wrong?
> 
> In any case, I have no space on the source to keep the latest common snapshot for too long, so I need a way to run normally if such snapshot exists or start from scratch otherwise. Is this possisble?
> 
> TIA

What you probably need is bookmarks: This way you can keep 0 snapshots on the source and still have it send incremental backups.

Note: Bookmarks are not supported yet, #23

**Comment by psy0rz** [comment]
i didnt read it correctly last time.

so destroying fails?

can you show a zfs list -r -t all backupdisk/vm/sql/disk0 ?

it seems there are other snapshots that are preventing the destroy.

**Comment by psy0rz** [comment]
(still holds are needed in this scenario to prevent common snapshot destruction)

**Comment by netfenceventu** [comment]
# zfs list -r -t all backup1/vm/sql/disk0
NAME                                          USED  AVAIL  REFER  MOUNTPOINT
backup1/vm/sql/disk0                          198G  4.61T  63.6G  -
backup1/vm/sql/disk0@auto_vm-20220425020000  2.15G      -  63.6G  -
backup1/vm/sql/disk0@auto_vm-20220426020000   164M      -  63.6G  -
backup1/vm/sql/disk0@auto_vm-20220426110009      0      -  63.6G  -

# zfs list -r -t all vm/sql/disk0
NAME                                  USED  AVAIL  REFER  MOUNTPOINT
vm/sql/disk0                          198G   149G  63.6G  -
vm/sql/disk0@auto_vm-20220426020000  2.20G      -  63.6G  -
vm/sql/disk0@auto_vm-20220427020000   193M      -  63.6G  -
vm/sql/disk0@auto_vm-20220427140000  14.3M      -  63.6G  -


As you can see we currently have a common snapshot (20220426020000) as last backup was taken yesterday; however in a few days we'll have none (and I have no space to hold more).

So what I'd like to achieve is:
_ if a common snapshot exists, just work normally;
_ otherwise destroy all target and start anew (or equivalent approach).



Thanks.

**Reply by psy0rz** [reply]
You just will have to use holds and use this as snapshot job:

> /usr/local/bin/zfs-autobackup --keep-source 1

This way you will have the absolute minimum amount of snapshots. 

Usually 2: The latest common snapshot and the newest snapshot. As soon as you run the backup the oldest snapshot will disappear and the latest one will become the common.

Holds are important to make use the snapshotter doesnt delete the common snapshot.

**Comment by netfenceventu** [comment]
Let me see if I understand correctly:

A) On the snapshot job, doesn't "--keep-source 1" keep the *last* snapshot? At least that's how I interpret the docs.
So how won't this prune the *last snapshot on the target* from the source (since it's not the latest on the source)?
Is that because of holds?
Will "--keep-source 1,1d2d" be ok also?

B) So I should remove --no-holds from the backup job, right?
Is there some more detailed docs on how holds are used?

C) I know of bookmarks (and use them elsewhere), but can zfs-autobackup use them?
They are not mentioned in the docs, but would help a lot in such a space-constrained situation.

TIA.

**Comment by psy0rz** [comment]
A. Yes, --keep-source would normally only keep the last snapshot, but the common snapshot is ALWAYS kept.

However, in snapshot-only mode we can only determine if something is a common snapshot by looking at the holds. (run it at least once without --no-holds so it will add the holds)

Any form of --keep-source is ok.

B. Yes. I'll add docs about it to the wiki. (i already created a seinfeld meme for it but forgot about writing it :)

C. They are not yet supported. See issue #23

**Reply by netfenceventu** [reply]
So I removed --no-holds from the backup script and set the snapshot script to use "--keep-source 1,1d2d".
Now I have:
# zfs list -t snap | grep vm/sql/disk0
vm/sql/disk0@auto_vm-20220504100002                     2.28G      -  64.5G  -
vm/sql/disk0@auto_vm-20220505020000                     1.17G      -  64.4G  -
vm/sql/disk0@auto_vm-20220505190000                     2.73G      -  64.8G  -

The first one has holds (as it's the common snapshot with target).
AFAICT 20220505190000 should be thinned with these settings, but it's not.

In fact now, even if I just use "--keep-source 1", I get:
# /usr/local/bin/zfs-autobackup --keep-source 1 auto_vm
! [Source] STDERR > cannot create snapshot 'vm/sql/disk0@auto_vm-20220506152539': out of space
! [Source] Command "zfs snapshot vm/sql/disk0@auto_vm-20220506152539" returned exit code 1 (valid codes: [0])
! Exception: Last command returned error

Do I understand thinning settings wrong?

TIA.

**Comment by psy0rz** [comment]
added a wiki page to explain all this a bit better: https://github.com/psy0rz/zfs_autobackup/wiki/Common-snapshots-and-holds
